### PR TITLE
Add type hints to some of the source code

### PR DESCRIPTION
**Pull request by andy-a-d-nguyen**
I added type hints to some of the source code as best as I could. I also followed recommendation from https://mypy.readthedocs.io/en/stable/type_narrowing.html to narrow down variables which are Optional types.

### CODE REVIEW

**Review by bblay — CHANGES_REQUESTED** [review]

File: source/fab/steps/compile_fortran.py [hunk 1]
```diff
@@ -385,7 +386,7 @@ def get_fortran_preprocessor():
     if not fpp:
         raise RuntimeError('no fortran preprocessor specified or discovered')
 
-    if '-P' not in fpp_flags:
+    if fpp_flags and '-P' not in fpp_flags:
```
[the remark this answers is not in the document]
Comment (andy-a-d-nguyen) [reply source/fab/steps/compile_fortran.py]: @bblay I reverted the if condition and used an assertion before the if condition like you suggested.

File: source/fab/steps/compile_fortran.py [hunk 2]
```diff
@@ -88,7 +88,7 @@ def __init__(self, compiler: Optional[str] = None, common_flags: Optional[List[s
         self.two_stage_flag = two_stage_flag
 
         # runtime, for child processes to read
-        self._stage = None
+        self._stage: int | None = None
```
[the remark this answers is not in the document]
Comment (andy-a-d-nguyen) [reply source/fab/steps/compile_fortran.py]: @bblay I changed the type hints to be compatible with Python 3.7

**Comment by bblay** [comment]
This looks good, Andy. Fab runs flake8 and mypy on its PRs. It just needs a little change to give us green ticks.
You can see the command used to run [flake8 here](https://github.com/metomi/fab/blob/master/.github/workflows/build.yml#L41) and [mypy here](https://github.com/metomi/fab/blob/master/.github/workflows/build.yml#L37).

I think case I think we just need to restore the Tuple import, as it's used elsewhere.

**Comment by andy-a-d-nguyen** [comment]
@bblay I re-added the Tuple import in psyclone.py and ran the flake8 linter on the source/ directory. Running the linter on the whole project seems to lint third-party libraries as well.

**Review by wilbertcs — APPROVED** [review]
Change was added as requested.

**Review by bblay — APPROVED** [review]

**Pull request merged**